sentinel-adapter/sentinel-apache-dubbo-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/dubbo/SentinelDubboProviderFilter.java

```diff
@@ -85,18 +105,62 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
             return result;
         } catch (BlockException e) {
             return DubboAdapterGlobalConfig.getProviderFallback().handle(invoker, invocation, e);
-        } catch (RpcException e) {
-            Tracer.traceEntry(e, interfaceEntry);
-            Tracer.traceEntry(e, methodEntry);
-            throw e;
         } finally {
             if (methodEntry != null) {
                 methodEntry.exit(1, invocation.getArguments());
             }
             if (interfaceEntry != null) {
                 interfaceEntry.exit();
             }
-            ContextUtil.exit();
+        }
+    }
+
+
+    private Result asyncInvoke(Invoker<?> invoker, Invocation invocation, String interfaceResourceName, String methodResourceName) {
+        LinkedList<EntryHolder> queue = new LinkedList<>();
+        try {
+            queue.push(new EntryHolder(
+                    SphU.asyncEntry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN), null));
+            queue.push(new EntryHolder(
+                    SphU.asyncEntry(methodResourceName, ResourceTypeConstants.COMMON_RPC,
+                            EntryType.IN, invocation.getArguments()), invocation.getArguments()));
+            Result result = invoker.invoke(invocation);
+            result.whenCompleteWithContext((r, throwable) -> {
+                Throwable error = throwable;
+                if (error == null) {
+                    error = Optional.ofNullable(r).map(Result::getException).orElse(null);
+                }
```

Comment:
It's elegant to use `Optional.ofNullable(r)` in some scenarios, but the side effect is that it'll create new `Optional` object, if `r` is not null.
Since here is a simple judgment scenario and the filter will be called every time the interface is called,
Will this be better? `error = r != null ? r.getException() : null;` 

Maybe in most cases, `r` is null if no exception occurred, I think it's acceptable using `Optional.ofNullable(r)`.

How do you think about it? @goodjava @sczyh30

Reply:
I think `error = r != null ? r.getException() : null` is more clear than using `optional` here